### PR TITLE
Spack package to add build options for CICE5 in esm1.6

### PR DESCRIPTION
**Pull request by anton-seaice**
@Harshula

When I try to use this package, it builds the wrong executable

I run 

`spack install cice5@git.access-esm1.6=access-esm1.6%intel@19.0.3.199`

and it builds the executable for access-cm2 ?

From the spack-build-out.txt file, I can see that spack is sending the wrong arguments to spack-build.sh :

```
==> cice5: Executing phase: 'build'
==> [2024-11-29-11:34:55.729889] '/g/data/tm70/as2285/spack0.22/spack/var/spack/environments/esm1_5/cice5/bld/spack-build.sh' 'access-cm2' '360x300' '12x1' '12'
```

Any ideas ?

### CODE REVIEW

**Comment by anton-seaice** [comment]
I think the issue is the use of strings as versions ...

Reading the spack docs, on [version comparison](https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison), I think 'access-esm1.6' is > 'access-cm2' because `e>c` ... Which is pretty confusing.

Reording the `with` statements could fix the problem, but is pretty confusing ... I am inclined to use variants instead ?

File: packages/cice5/package.py [hunk 1]
```diff
@@ -14,9 +14,11 @@ class Cice5(MakefilePackage):
     homepage = "https://www.access-nri.org.au"
     git = "https://github.com/ACCESS-NRI/cice5.git"
 
-    maintainers = ["harshula"]
+    maintainers = ["harshula", "anton-seaice"]
 
-    version("master", branch="master")
+    version("master", branch="master", preferred=True)
+    version("access-esm1.6", branch="access-esm1.6")
+    version("access-cm2", branch="access-cm2")
```
Comment (harshula) [inline packages/cice5/package.py:21]: I don't see an `access-cm2` branch in the cice5 repository.

File: packages/cice5/package.py [hunk 2]
```diff
@@ -25,45 +27,64 @@ class Cice5(MakefilePackage):
...
-    depends_on("parallelio~pnetcdf~timing~shared")
+    
     depends_on("datetime-fortran")
     depends_on("oasis3-mct+deterministic", when="+deterministic")
     depends_on("oasis3-mct~deterministic", when="~deterministic")
-    depends_on("libaccessom2+deterministic", when="+deterministic")
-    depends_on("libaccessom2~deterministic", when="~deterministic")
+
+
+    with when("@master"):
```
Comment (harshula) [inline packages/cice5/package.py:36]: Use what's done in MOM5: `with when("@:access-esm0,access-esm2:"):`

File: packages/cice5/package.py [hunk 3]
```diff
@@ -25,45 +27,64 @@ class Cice5(MakefilePackage):
...
+
+        __targets[1682]["driver"] = "auscom"
+        __targets[1682]["grid"] = "3600x2700"
+        __targets[1682]["blocks"] = "200x180" 
+
+    with when("@access-cm2:"):
+        __targets = {12: {}} 
+        __targets[12]["driver"] = "access-cm2"
+        __targets[12]["grid"] = "360x300"
+        __targets[12]["blocks"] = "12x1"
```
Comment (harshula) [inline packages/cice5/package.py:73]: Remove until access-cm2 is ready.

File: packages/cice5/spack-build.sh [hunk 4]
```diff
@@ -37,26 +37,37 @@ endif
 ### Specialty code
 setenv CAM_ICE  no        # set to yes for CAM runs (single column)
 setenv SHRDIR   csm_share # location of CCSM shared code
-setenv IO_TYPE  pio       # set to none if netcdf library is unavailable
+setenv IO_TYPE  netcdf    # set to none if netcdf library is unavailable
```
Comment (harshula) [inline packages/cice5/spack-build.sh:40]: Maybe remove ?

File: packages/cice5/spack-build.sh [hunk 5]
```diff
@@ -37,26 +37,37 @@ endif
...
+    setenv NICELYR    4     #4 for standard multi-layer ice (ktherm=1)
+else #driver = auscom
+    setenv DRVDIR $driver
     setenv ACCESS   no
+    setenv IO_TYPE  pio
+    setenv NICELYR    4       # number of vertical layers in the ice
+    ### The version of an executable can be found with the following
+    ### command: strings <executable> | grep 'CICE_VERSION='
+    # set version='202301'
+    # sed -e "s/{CICE_VERSION}/$version/g" $SRCDIR/drivers/$driver/version.F90.template > $SRCDIR/drivers/$driver/version_mod.F90
```
Comment (anton-seaice) [inline packages/cice5/spack-build.sh:68]:
```suggestion
    set version='202301'
    sed -e "s/{CICE_VERSION}/$version/g" $SRCDIR/drivers/$driver/version.F90.template > $SRCDIR/drivers/$driver/version_mod.F90
```

Comment (harshula) [inline packages/cice5/spack-build.sh:68]: Probably needed by ACCESS-OM2.

File: packages/cice5/package.py [hunk 6]
```diff
@@ -25,45 +27,64 @@ class Cice5(MakefilePackage):
...
+
+    with when("@=access-esm1.6"):
+        # possibly we should revert to netcdf/4.6.3
```
Comment (harshula) [inline packages/cice5/package.py:76]: Remove ?

File: packages/cice5/package.py [hunk 7]
```diff
@@ -97,14 +118,22 @@ def edit(self, spec, prefix):
...
+        if self.spec.satisfies("master"):
+            lstr = self.make_linker_args(spec, "parallelio", "-lpiof -lpioc")
+            ideps.extend(["libaccessom2"])
+
+            # NOTE: The order of the libraries matter during the linking step!
+            # NOTE: datetime-fortran is a dependency of libaccessom2.
+            ldeps = ["oasis3-mct", "libaccessom2", "netcdf-c", "netcdf-fortran", "datetime-fortran"]
+
         libs = " ".join([lstr] + [self.get_linker_args(spec, d) for d in ldeps])
+        libs = " ".join([self.get_linker_args(spec, d) for d in ldeps])
```
Comment (harshula) [inline packages/cice5/package.py:136]: Follow the pattern in the MOM5 SPR: https://github.com/ACCESS-NRI/spack-packages/blob/ff2fbc0a9680118a805a0577edbe03008c5346ae/packages/mom5/package.py#L67-L87

**Comment by anton-seaice** [comment]
@harshula - I made the suggested changes, but it didn't generate the __targets correctly

I moved the __targets to the build() step and now it works

For access-om2, see the pre-release for this version of spack-packages:

**Comment by anton-seaice** [comment]
New build is bitwise reproducible with the old, see https://github.com/ACCESS-NRI/access-om2-configs/pull/151#issuecomment-2518853690

**Comment by anton-seaice** [comment]
Closes https://github.com/ACCESS-NRI/spack-packages/issues/156

**Comment by harshula** [comment]
> @harshula - I made the suggested changes, but it didn't generate the __targets correctly
> I moved the __targets to the build() step and now it works

Did it generate the targets correctly before you made the suggested changes?

**Comment by anton-seaice** [comment]
> > @harshula - I made the suggested changes, but it didn't generate the __targets correctly
> > I moved the __targets to the build() step and now it works
> 
> Did it generate the targets correctly before you made the suggested changes?

No - see first comment in the PR.

**Comment by harshula** [comment]
Thanks for the clarification. It is best to avoid declaring member variables (`self.__targets = __targets`) from a member function. Ideally, the `targets` data structure should be declared after the class declaration. I'll look into whether `with when()` allows variable declaration/definition. This might be a `with when()` limitation.

**Comment by anton-seaice** [comment]
Ok thanks. The targets are only used in the build function, so it could just me a local
Variable instead of a member variable

**Comment by harshula** [comment]
No, they are also used in the `install` member function.

**Comment by anton-seaice** [comment]
Hi @ofa001 - just confirming a couple of science build settings for CICE5 in access-esm1.6


I have set these two lines:

Zero layer ice:
`  setenv NICELYR  1     #1 for ktherm=0, zero-layer thermodynamics`

No BGC tracers:
`  setenv TRBGCS   0`

Is 5 thickness categories and 1 snow layer correct ? Do these two lines need changing?:

```
setenv NSNWLYR    1       # number of vertical layers in the snow
setenv NICECAT    5       # number of ice thickness categories
```

This is set to build with 12 tasks (to match esm1.5) but also will build a version with 16 tasks if needed. (Obviously these can be refined in the future).

File: packages/cice5/package.py (hunk 7 above)
Comment (anton-seaice) [reply packages/cice5/package.py]: Resolved in https://github.com/ACCESS-NRI/spack-packages/pull/179

File: packages/cice5/spack-build.sh (hunk 5 above)
Comment (anton-seaice) [reply packages/cice5/spack-build.sh]: Done

File: packages/cice5/spack-build.sh (hunk 4 above)
Comment (anton-seaice) [reply packages/cice5/spack-build.sh]: Done

File: packages/cice5/package.py (hunk 3 above)
Comment (anton-seaice) [reply packages/cice5/package.py]: Done

File: packages/cice5/package.py (hunk 2 above)
Comment (anton-seaice) [reply packages/cice5/package.py]: resolved in https://github.com/ACCESS-NRI/spack-packages/pull/179

**Comment by anton-seaice** [comment]
@harshula this is ready for review.

Produced two binaries expected:
```
$ ls /g/data/tm70/as2285/spack0.22/restricted/ukmo/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/cice5-git.access-esm1.6_access-esm1.6-nipftaviflymvjbwuxhzqycxpfucyexn/bin/
cice_access-esm1.6_360x300_12x1_12p.exe  cice_access-esm1.6_360x300_8x2_16p.exe

**Comment by ofa001** [comment]
HI @anton-seaice @harshula,  @DaveBi is holding off finishing the coupling the work as I understand to the New Year whilst he gets the cice4 version with the icebergs linked up with Pearce's Generic tracer for the test run's in Rachel's table.  So whilst your settings for cice5  ice thickness categoriex and snow ice layers are thin Anton,  it wont run, not sure it can pick tha tit doesnt have the correct links in to the UM in this spack set up as I am not sure how sophisticated it is.  We can talk about it at the mini-hackathon tomorrow if you are around.

**Comment by anton-seaice** [comment]
Thanks @ofa001 - this is just build infrastructure, I am just making sure its ready / available so we can build the model so the cice code can be updated.



> So whilst your settings for cice5 ice thickness categoriex and snow ice layers are thin Anton

What did you mean here ? We can change these compile time options later ofcourse, but its good if they are close now.

**Comment by ofa001** [comment]
Sorry  @anton-seaice, I was distracted,  I had someone here fixing my aircon, I meant "fine"

**Pull request merged**